File: Data Structures & Algorithms/coin-change-ii/submission-3.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        coins.sort()
        n=len(coins)
        d={}
        def dfs(i,a):
            if (i,a) in d:
                return d[(i,a)]
            if a==0:
                return 1
            if i>=n:
                return 0
            res=0
            if a>=coins[i]:
                res = dfs(i+1,a)
                res+=dfs(i,a-coins[i])
            d[(i,a)]=res
            return res
        return dfs(0,amount)        

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-3.java
class Solution {
    Map<Character,String> D =new HashMap<>(Map.of(
        '2',"abc",
        '3',"def",
        '4',"ghi",
        '5',"jkl",
        '6',"mno",
        '7',"pqrs",
        '8',"tuv",
        '9',"wxyz"
    ));
    List<String> res = new ArrayList<>();
    StringBuilder cur = new StringBuilder();
    public void dfs(String digits,int i , int k){
        if( k == D.get(digits.charAt(i)).length()  ) return ;
        cur.append(D.get(digits.charAt(i)).charAt(k) );
        if( i == digits.length()-1){
            res.add(cur.toString());
            cur.deleteCharAt(cur.length() - 1);
            dfs(digits,i,k+1);
        }
        else{
            dfs(digits,i+1,0);
            cur.deleteCharAt(cur.length() - 1); 
            dfs(digits,i,k+1);
        }
    }
    public List<String> letterCombinations(String digits) {
        if(digits.length()==0) return new ArrayList();
        dfs(digits,0,0);
        return this.res;
    }
}
